File: FMELibrary/Competition.cs
namespace FMELibrary
{
    /// <summary>
    /// Represents a football competition (league, cup, etc.) with all its properties.
    /// </summary>
    public class Competition
    {
        /// <summary>
        /// Gets or sets the competition identifier.
        /// </summary>
        public short Id { get; set; }

        /// <summary>
        /// Gets or sets the unique identifier for the competition.
        /// </summary>
        public int Uid { get; set; }

        /// <summary>
        /// Gets or sets the full name of the competition.
        /// </summary>
        public string FullName { get; set; }

        /// <summary>
        /// Gets or sets an unknown byte value.
        /// </summary>
        public byte Unknown1 { get; set; }

        /// <summary>
        /// Gets or sets the short name of the competition.
        /// </summary>
        public string ShortName { get; set; }

        /// <summary>
        /// Gets or sets an unknown byte value.
        /// </summary>
        public byte Unknown2 { get; set; }

        /// <summary>
        /// Gets or sets the code name of the competition.
        /// </summary>
        public string CodeName { get; set; }

        /// <summary>
        /// Gets or sets the type of competition.
        /// </summary>
        public byte Type { get; set; }

        /// <summary>
        /// Gets or sets the continent identifier.
        /// </summary>
        public short ContinentId { get; set; }

        /// <summary>
        /// Gets or sets the nation identifier.
        /// </summary>
        public short NationId { get; set; }

        /// <summary>
        /// Gets or sets the first color value.
        /// </summary>
        public short Color1 { get; set; }

        /// <summary>
        /// Gets or sets the second color value.
        /// </summary>
        public short Color2 { get; set; }

        /// <summary>
        /// Gets or sets the competition's reputation.
        /// </summary>
        public short Reputation { get; set; }

        /// <summary>
        /// Gets or sets the competition level.
        /// </summary>
        public byte Level { get; set; }

        /// <summary>
        /// Gets or sets the main competition identifier.
        /// </summary>
        public short MainComp { get; set; }

        /// <summary>
        /// Gets or sets the array of qualifier data (8 bytes each).
        /// </summary>
        public byte[][] Qualifiers { get; set; }

        /// <summary>
        /// Gets or sets the first ranking value.
        /// </summary>
        public int Rank1 { get; set; }

        /// <summary>
        /// Gets or sets the second ranking value.
        /// </summary>
        public int Rank2 { get; set; }

        /// <summary>
        /// Gets or sets the third ranking value.
        /// </summary>
        public int Rank3 { get; set; }

        /// <summary>
        /// Gets or sets the first year value.
        /// </summary>
        public short Year1 { get; set; }

        /// <summary>
        /// Gets or sets the second year value.
        /// </summary>
        public short Year2 { get; set; }

        /// <summary>
        /// Gets or sets the third year value.
        /// </summary>
        public short Year3 { get; set; }

        /// <summary>
        /// Gets or sets an unknown byte value.
        /// </summary>
        public byte Unknown3 { get; set; }

        /// <summary>
        /// Gets or sets an unknown byte value.
        /// </summary>
        public byte Unknown4 { get; set; }

        #region Extra

        /// <summary>
        /// Gets or sets the nation name (extra/computed field).
        /// </summary>
        public string Nation { get; set; } = string.Empty;

        #endregion

        /// <summary>
        /// Initializes a new instance of the <see cref="Competition"/> class by reading from a binary reader.
        /// </summary>
        /// <param name="reader">The binary reader containing the competition data.</param>
        public Competition(BinaryReader reader)
        {
            Id = reader.ReadInt16();
            Uid = reader.ReadInt32();

            FullName = reader.ReadStringEx();
            Unknown1 = reader.ReadByte();
            ShortName = reader.ReadStringEx();
            Unknown2 = reader.ReadByte();
            CodeName = reader.ReadStringEx();

            Type = reader.ReadByte();
            ContinentId = reader.ReadInt16();
            NationId = reader.ReadInt16();
            Color1 = reader.ReadInt16();
            Color2 = reader.ReadInt16();
            Reputation = reader.ReadInt16();
            Level = reader.ReadByte();
            MainComp = reader.ReadInt16();

            Qualifiers = new byte[reader.ReadInt32()][];
            for (int i = 0; i < Qualifiers.Length; i++)
            {
                Qualifiers[i] = reader.ReadBytes(8);
            }

            Rank1 = reader.ReadInt32();
            Rank2 = reader.ReadInt32();
            Rank3 = reader.ReadInt32();
            Year1 = reader.ReadInt16();
            Year2 = reader.ReadInt16();
            Year3 = reader.ReadInt16();

            Unknown3 = reader.ReadByte();
            Unknown4 = reader.ReadByte();
        }

        /// <summary>
        /// Converts the competition data to a byte array.
        /// </summary>
        /// <returns>A byte array representing the serialized competition data.</returns>
        public byte[] ToBytes()
        {
            using var stream = new MemoryStream();
            using var writer = new BinaryWriter(stream);
            Write(writer);
            return stream.ToArray();
        }

        /// <summary>
        /// Writes the competition data to the specified binary writer.
        /// </summary>
        /// <param name="writer">The binary writer to write the competition data to.</param>
        public void Write(BinaryWriter writer)
        {
            writer.WriteEx(Id);
            writer.WriteEx(Uid);

            writer.WriteEx(FullName);
            writer.WriteEx(Unknown1);
            writer.WriteEx(ShortName);
            writer.WriteEx(Unknown2);
            writer.WriteEx(CodeName);

            writer.WriteEx(Type);
            writer.WriteEx(ContinentId);
            writer.WriteEx(NationId);
            writer.WriteEx(Color1);
            writer.WriteEx(Color2);
            writer.WriteEx(Reputation);
            writer.WriteEx(Level);
            writer.WriteEx(MainComp);

            writer.WriteEx(Qualifiers.Length);
            for (int i = 0; i < Qualifiers.Length; i++)
            {
                writer.WriteEx(Qualifiers[i]);
            }

            writer.WriteEx(Rank1);
            writer.WriteEx(Rank2);
            writer.WriteEx(Rank3);
            writer.WriteEx(Year1);
            writer.WriteEx(Year2);
            writer.WriteEx(Year3);
            writer.WriteEx(Unknown3);
            writer.WriteEx(Unknown4);
        }

        /// <summary>
        /// Returns a string representation of the competition.
        /// </summary>
        /// <returns>A string containing the competition's ID, UID, and full name.</returns>
        public override string ToString()
        {
            return $"{Id} - {Uid} - {FullName}";
        }
    }
}

File: FMELibrary.js/Competition.ts
import { BinaryParser } from './BinaryParser';

/**
 * Represents a football competition (league, cup, etc.) with all its properties.
 */
export class Competition {
  /** Gets or sets the competition identifier. */
  id: number = 0;

  /** Gets or sets the unique identifier for the competition. */
  uid: number = 0;

  /** Gets or sets the full name of the competition. */
  fullName: string = '';

  /** Gets or sets an unknown byte value. */
  unknown1: number = 0;

  /** Gets or sets the short name of the competition. */
  shortName: string = '';

  /** Gets or sets an unknown byte value. */
  unknown2: number = 0;

  /** Gets or sets the code name of the competition. */
  codeName: string = '';

  /** Gets or sets the type of competition. */
  type: number = 0;

  /** Gets or sets the continent identifier. */
  continentId: number = 0;

  /** Gets or sets the nation identifier. */
  nationId: number = 0;

  /** Gets or sets the first color value. */
  color1: number = 0;

  /** Gets or sets the second color value. */
  color2: number = 0;

  /** Gets or sets the competition's reputation. */
  reputation: number = 0;

  /** Gets or sets the competition level. */
  level: number = 0;

  /** Gets or sets the main competition identifier. */
  mainComp: number = 0;

  /** Gets or sets the array of qualifier data (8 bytes each). */
  qualifiers: Buffer[] = [];

  /** Gets or sets the first ranking value. */
  rank1: number = 0;

  /** Gets or sets the second ranking value. */
  rank2: number = 0;

  /** Gets or sets the third ranking value. */
  rank3: number = 0;

  /** Gets or sets the first year value. */
  year1: number = 0;

  /** Gets or sets the second year value. */
  year2: number = 0;

  /** Gets or sets the third year value. */
  year3: number = 0;

  /** Gets or sets an unknown byte value. */
  unknown3: number = 0;

  /** Gets or sets an unknown byte value. */
  unknown4: number = 0;

  // Extra fields
  /** Gets or sets the nation name (extra/computed field). */
  nation: string = '';

  /**
   * Initializes a new instance of the Competition class by reading from a buffer.
   * @param buffer - The buffer containing the competition data
   * @param offset - The current offset in the buffer
* @returns An object containing the Competition instance and the new offset
   */
  static fromBuffer(buffer: Buffer, offset: number): { competition: Competition; offset: number } {
    const competition = new Competition();

    competition.id = buffer.readInt16LE(offset);
    offset += 2;

    competition.uid = buffer.readInt32LE(offset);
    offset += 4;

    const fullNameResult = BinaryParser.readStringEx(buffer, offset);
    competition.fullName = fullNameResult.value;
    offset = fullNameResult.offset;

    competition.unknown1 = buffer.readUInt8(offset);
    offset += 1;

    const shortNameResult = BinaryParser.readStringEx(buffer, offset);
    competition.shortName = shortNameResult.value;
    offset = shortNameResult.offset;

    competition.unknown2 = buffer.readUInt8(offset);
    offset += 1;

    const codeNameResult = BinaryParser.readStringEx(buffer, offset);
    competition.codeName = codeNameResult.value;
    offset = codeNameResult.offset;

    competition.type = buffer.readUInt8(offset);
    offset += 1;

    competition.continentId = buffer.readInt16LE(offset);
    offset += 2;

    competition.nationId = buffer.readInt16LE(offset);
    offset += 2;

    competition.color1 = buffer.readInt16LE(offset);
    offset += 2;

    competition.color2 = buffer.readInt16LE(offset);
    offset += 2;

    competition.reputation = buffer.readInt16LE(offset);
    offset += 2;

    competition.level = buffer.readUInt8(offset);
    offset += 1;

    competition.mainComp = buffer.readInt16LE(offset);
    offset += 2;

    const qualifiersCount = buffer.readInt32LE(offset);
    offset += 4;

    competition.qualifiers = [];
    for (let i = 0; i < qualifiersCount; i++) {
      competition.qualifiers.push(buffer.slice(offset, offset + 8));
      offset += 8;
    }

    competition.rank1 = buffer.readInt32LE(offset);
    offset += 4;

    competition.rank2 = buffer.readInt32LE(offset);
    offset += 4;

    competition.rank3 = buffer.readInt32LE(offset);
    offset += 4;

    competition.year1 = buffer.readInt16LE(offset);
    offset += 2;

    competition.year2 = buffer.readInt16LE(offset);
    offset += 2;

    competition.year3 = buffer.readInt16LE(offset);
    offset += 2;

    competition.unknown3 = buffer.readUInt8(offset);
    offset += 1;

    competition.unknown4 = buffer.readUInt8(offset);
    offset += 1;

    return { competition, offset };
  }

  /**
* Converts the competition data to a buffer.
   * @returns A buffer representing the serialized competition data
   */
  toBuffer(): Buffer {
    let size = 2 + 4; // id + uid
    size += 4 + Buffer.byteLength(this.fullName, 'utf-8'); // fullName
    size += 1; // unknown1
    size += 4 + Buffer.byteLength(this.shortName, 'utf-8'); // shortName
    size += 1; // unknown2
    size += 4 + Buffer.byteLength(this.codeName, 'utf-8'); // codeName
    size += 1 + 2 + 2 + 2 + 2 + 2 + 1 + 2; // type to mainComp
    size += 4 + (this.qualifiers.length * 8); // qualifiers
    size += 4 + 4 + 4 + 2 + 2 + 2 + 1 + 1; // ranks, years, unknowns

    const buffer = Buffer.alloc(size);
    let offset = 0;

    buffer.writeInt16LE(this.id, offset);
    offset += 2;

    buffer.writeInt32LE(this.uid, offset);
    offset += 4;

    offset = BinaryParser.writeEx(this.fullName, buffer, offset);
    offset = BinaryParser.writeByte(this.unknown1, buffer, offset);
    offset = BinaryParser.writeEx(this.shortName, buffer, offset);
    offset = BinaryParser.writeByte(this.unknown2, buffer, offset);
    offset = BinaryParser.writeEx(this.codeName, buffer, offset);

    buffer.writeUInt8(this.type, offset);
    offset += 1;

    buffer.writeInt16LE(this.continentId, offset);
    offset += 2;

    buffer.writeInt16LE(this.nationId, offset);
    offset += 2;

    buffer.writeInt16LE(this.color1, offset);
    offset += 2;

    buffer.writeInt16LE(this.color2, offset);
    offset += 2;

    buffer.writeInt16LE(this.reputation, offset);
    offset += 2;

    buffer.writeUInt8(this.level, offset);
    offset += 1;

    buffer.writeInt16LE(this.mainComp, offset);
    offset += 2;

    buffer.writeInt32LE(this.qualifiers.length, offset);
    offset += 4;

    for (const qualifier of this.qualifiers) {
      qualifier.copy(buffer, offset);
      offset += 8;
    }

    buffer.writeInt32LE(this.rank1, offset);
    offset += 4;

    buffer.writeInt32LE(this.rank2, offset);
    offset += 4;

    buffer.writeInt32LE(this.rank3, offset);
    offset += 4;

    buffer.writeInt16LE(this.year1, offset);
    offset += 2;

    buffer.writeInt16LE(this.year2, offset);
    offset += 2;

    buffer.writeInt16LE(this.year3, offset);
    offset += 2;

    buffer.writeUInt8(this.unknown3, offset);
    offset += 1;

    buffer.writeUInt8(this.unknown4, offset);
    offset += 1;

    return buffer.slice(0, offset);
  }

  /**
   * Returns a string representation of the competition.
   * @returns A string containing the competition's ID, UID, and full name
   */
  toString(): string {
    return `${this.id} - ${this.uid} - ${this.fullName}`;
  }
}
